core/src/main/java/org/jdbi/v3/core/collector/MapCollectorFactory.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.collector;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.reflect.TypeParameter;
+import com.google.common.reflect.TypeToken;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+
+import static org.jdbi.v3.core.generic.GenericTypes.getErasedType;
+
+class MapCollectorFactory implements CollectorFactory {
+    private static final TypeVariable<Class<Map>> KEY_PARAM;
+    private static final TypeVariable<Class<Map>> VALUE_PARAM;
+
+    static {
+        TypeVariable<Class<Map>>[] mapParams = Map.class.getTypeParameters();
+        KEY_PARAM = mapParams[0];
+        VALUE_PARAM = mapParams[1];
+    }
+
+    private static final Map<Class<?>, Collector<?,?,?>> collectors = ImmutableMap.<Class<?>, Collector<?,?,?>>builder()
```

Comment:
I guess it makes sense to declare this variable in the upper-case as a constant, because it's effectively immutable.